KAFKA-8813: Refresh log config if it's updated before initialization

soondenana opened the pull request:
A partition log in initialized in following steps:

1. Fetch log config from ZK
2. Call LogManager.getOrCreateLog which creates the Log object, then
3. Registers the Log object

Step #3 enables Configuration update thread to deliver configuration
updates to the log. But if any update arrives between step #1 and #3
then that update is missed. It breaks following use case:

1. Create a topic with default configuration, and immediately after that
2. Update the configuration of topic

There is a race condition here and in random cases update made in
seocond step will get dropped.

This change fixes it by tracking updates arriving between step #1 and #3
Once a Partition is done initialzing log, it checks if it has missed any
update. If yes, then the configuration is read from ZK again.

Added unit tests to make sure a dirty configuration is refreshed. Tested
on local cluster to make sure that topic configuration and updates are
handled correctly.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
KAFKA-8813: Refresh log config if it's updated before initialization

A partition log in initialized in following steps:

1. Fetch log config from ZK
2. Call LogManager.getOrCreateLog which creates the Log object, then
3. Registers the Log object

Step #3 enables Configuration update thread to deliver configuration
updates to the log. But if any update arrives between step #1 and #3
then that update is missed. It breaks following use case:

1. Create a topic with default configuration, and immediately after that
2. Update the configuration of topic

There is a race condition here and in random cases update made in
seocond step will get dropped.

This change fixes it by tracking updates arriving between step #1 and #3
Once a Partition is done initialzing log, it checks if it has missed any
update. If yes, then the configuration is read from ZK again.

Added unit tests to make sure a dirty configuration is refreshed. Tested
on local cluster to make sure that topic configuration and updates are
handled correctly.

soondenana commented:
cc: @hachikuji, @jsancio, @mumrah 

Maybe useful to review the change by hiding whitespace changes (put `?w=1` to the end of url or use the UI)
![HideWhitespace](https://user-images.githubusercontent.com/50422828/64392264-cf6b3d80-d000-11e9-9dd0-bf3bee37e822.png)

soondenana commented:
Ran create topic, change config in a loop (ran 100 times) and made sure that configs update correctly. Here is the bash command I used to create and change topic:

```bash
for idx in `seq 100`; do 
    bin/kafka-topics.sh --create --bootstrap-server localhost:9092 --replication-factor 3 --partitions 3 --topic test-topic-${idx}
    bin/kafka-configs.sh --alter --entity-type topics --entity-name test-topic-${idx} --add-config 'min.insync.replicas=2' --zookeeper localhost:2181
done
```

Here is one to check the configuration:
```bash
for idx in `seq 100`; do 
    bin/kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic test-topic-${idx}
done 2> /dev/null | grep min.insync
```

Here is to clean things up at the end:
```bash
for idx in `seq 100`; do 
    bin/kafka-topics.sh --delete --bootstrap-server localhost:9092  --topic test-topic-${idx}
done
```

soondenana commented:
Tested again with 500 topic creation followed by update (instead of 100 in previous comment). Checked to make sure topic configuration is sane.

hachikuji commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -310,17 +310,28 @@ class Partition(val topicPartition: TopicPartition,
-      0L
+  // Visible for testing
+  private[cluster] def createLog(replicaId: Int, isNew: Boolean, isFutureReplica: Boolean, offsetCheckpoints: OffsetCheckpoints): Log = {
+    def logConfig: LogConfig = {
```
nit: how about `fetchLogConfig()`?

soondenana replied:
Done.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +669,47 @@ class LogManager(logDirs: Seq[File],
+                              log: Option[Log],
+                              config: => LogConfig): Unit = configUpdateLock synchronized {
+    if (partitionsInitializing(topicPartition)) {
+      log.foreach(_.config = config)
```
Seems undesirable to be holding the lock while fetching the topic config. Would it be possible to let this return a boolean indicating whether the config needs to be reinitialized rather than doing it in place?

Another thing I am considering is whether we can protect updates to the config using the version from ZK. Basically only override the config if we have a higher version. Then we don't have to worry as much about local ordering of updates.

soondenana replied:
Updated to use ConcurrentHashMap. Now the code checks and refreshes config in a loop to take care of race.

hachikuji reviewed (commented):
Thanks for the patch. Left a couple initial comments.

soondenana committed:
Take care of review comments

This change takes care of issue where we were holding lock while talking
to ZK. The change switches to use a concurrent hash map and retries
config load operation if it has been changed since it got refreshed.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +666,56 @@ class LogManager(logDirs: Seq[File],
+   */
+  def finishedInitializingLog(topicPartition: TopicPartition,
+                              maybeLog: Option[Log],
+                              config: => LogConfig): Unit = {
```
nit: I think it would be clearer to turn this into a normal function with a name like `fetchConfig`

soondenana replied:
Done.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -82,6 +82,13 @@ class LogManager(logDirs: Seq[File],
+  /**
+   * Lock object to protect modification to partitionsInitializing variable.
+   */
+  private val configUpdateLock = new Object
```
Would it be more appropriate to call this `initializationLock` or something like that. It does not protect normal config updates.

soondenana replied:
Got rid of lock as it isn't adding any value if we don't use `while` loop below. The map is concurrent hash map, so that handles concurrent updates to the key/value.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +666,56 @@ class LogManager(logDirs: Seq[File],
+    /*
+     * Keep on reloading config as long as it keeps changing. A topic or broker config change
+     * event (in ConfigHandler::processConfigChanges or DynamicLogConfig::reconfigure) can
+     * arrive while we are loading config. If that happens reload the configuration.
```
Probably worth mentioning that we are not holding the lock because fetching the config can be somewhat expensive.

soondenana replied:
No longer applicable as lock is removed.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -955,9 +1012,9 @@ class LogManager(logDirs: Seq[File],
-      topicPartition.topic() == topic
-    }.map { case (_, log) => log }
+    (currentLogs.toList ++ futureLogs.toList).collect {
+      case (topicPartition, log) if topicPartition.topic() == topic => log
```
nit: unneeded parenthesis in `topic()`

soondenana replied:
Done.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +666,56 @@ class LogManager(logDirs: Seq[File],
+     */
+    while (partitionsInitializing(topicPartition)) {
+      partitionsInitializing(topicPartition) = false;
+      maybeLog.foreach(_.config = config)
```
Should we use `Log.updateConfig`?

soondenana replied:
Done. Passed an empty set for `updatedKeys`.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +666,56 @@ class LogManager(logDirs: Seq[File],
+     * event (in ConfigHandler::processConfigChanges or DynamicLogConfig::reconfigure) can
+     * arrive while we are loading config. If that happens reload the configuration.
+     */
+    while (partitionsInitializing(topicPartition)) {
```
I think I understand the case we're trying to protect, but possibly it borders on overkill because I'm not sure how we can guarantee a correct ordering of updates in all cases. For example, in `TopicConfigHandler. updateLogConfig`, we have a similar race condition where we fetch the config, then update it.

soondenana replied:
Done. Replaced `while` with `if`.

hachikuji commented on core/src/main/scala/kafka/server/ConfigHandler.scala:
```diff
@@ -54,16 +54,20 @@ class TopicConfigHandler(private val logManager: LogManager, kafkaConfig: KafkaC
-      val props = new Properties()
-      topicConfig.asScala.foreach { case (key, value) =>
-        if (!configNamesToExclude.contains(key)) props.put(key, value)
+    def updateLogConfig: Unit = {
```
Inner defs are nice when they are small. Here I wonder if we should just turn this into a `private def` instead so we can avoid cluttering the logic. Same for `updateThrottledList`.

Similar comment in `DynamicBrokerConfig`.

soondenana replied:
Done.

hachikuji reviewed (commented):
Sorry for the delay. Left a few more comments.

soondenana committed:
Take care of review comments : Part 2

soondenana committed:
Remove updatedKeys param and related code from Log::updateConfig

Removed the param and checked the config value directly instead.

soondenana committed:
fix scala 2.11 compilation error

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -82,6 +82,9 @@ class LogManager(logDirs: Seq[File],
   @volatile private var numRecoveryThreadsPerDataDir = recoveryThreadsPerDataDir
 
+  // Visible for testing
+  private[log] val partitionsInitializing = new ConcurrentHashMap[TopicPartition, Boolean]().asScala
```
Can you add a comment indicating that the value indicates that config was updated while the partition was being initialized?

soondenana replied:
Done.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -659,6 +662,48 @@ class LogManager(logDirs: Seq[File],
+   */
+  def brokerConfigUpdated(): Unit = {
+    partitionsInitializing.foreach {
+      case (topicPartition, _) => partitionsInitializing(topicPartition) = true
```
We might need to be a little careful since we no longer have the lock. It seems possible that `finishedInitializingLog` could be called concurrently. In that case, should we use a method like `replace` so that we atomically update the value only if it its current value is `false`? Similarly in `topicConfigUpdated`.

soondenana replied:
Thanks for catching this. Fixed.

hachikuji reviewed (commented):
Just a couple more comments.

soondenana committed:
Use Map.replace to add entry only if it is present

We took out locks from the config update code, which exposed a condition
where log entry can stay in the map if it was removed while an update
was already in progress. This change uses the `replace` method of the
Map to make sure that we don't insert an entry if it is not already
present.

Ran unit tests and they pass after this change.

hachikuji reviewed (approved):
LGTM. Thanks for the patch.

soondenana commented:
retest this please
